create a filter to suppress violations in git ignored files

lpapazow opened the pull request:
Provides a solution for #255.

With the addition of this filter, checkstyle violations for git ignored files will not be logged.

The current solution for #255 is to add supressions for each check in smarthome checkstyle/suppressions.xml file. However, the suppressions file needs to be updated every time in every repo when a new check is added or renamed. Having this filter should allow this to be avoided.

Signed-off-by: Lyubomir V. Papazov <lpapazow@gmail.com>

martinvw commented on sat-plugin/src/main/resources/rulesets/checkstyle/rules.xml:
```diff
@@ -12,6 +12,7 @@
  -->
 
 <module name="Checker">
+  <module name="org.openhab.tools.analysis.checkstyle.filters.GitIgnoredFilesFilter"/>
```
Formatting looks of

martinvw replied:
Nevermind

martinvw commented on custom-checks/checkstyle/src/main/java/org/openhab/tools/analysis/checkstyle/filters/GitIgnoredFilesFilter.java:
```diff
@@ -0,0 +1,46 @@
+
+        Path pathToGitIgnore = pathToFolder.resolve(GITIGNORE_FILE_NAME);
+        File gitIgnoreFile = pathToGitIgnore.toFile();
+        if (gitIgnoreFile.exists()) {
```
```java
return !gitIgnoreFile.exists()

lpapazow commented:
@martinvw, I changed the logic completely, can you please have another look

> I do not yet see how this would work, is the service offered so good that it can handle all the syntax of gitignore files, including question marks, *, partial and full paths?

You are right, it wouldn't have been able to handle all the syntax of .gitignore files just using the .gitignore path. Instead, I now added the `jgit` library, which can provide a list of all the files that are ignored by git.

dstivanov commented on custom-checks/checkstyle/src/main/java/org/openhab/tools/analysis/checkstyle/filters/GitIgnoredFilesFilter.java:
```diff
@@ -0,0 +1,67 @@
+ *         git. In order to activate the filter, A module with the full class
+ *         name should be added within the Checker module in the rules file.
+ */
+public class GitIgnoredFilesFilter extends SuppressionFilter {
```
Is is possible to add some unit tests for this filter?

dstivanov commented on pom.xml:
```diff
@@ -66,6 +66,7 @@
     <maven.surefire.plugin.version>2.12.4</maven.surefire.plugin.version>
     <jetty.server.version>9.3.14.v20161028</jetty.server.version>
     <saxon.version>9.1.0.8</saxon.version>
+    <jgit.version>5.1.1.201809181055-r</jgit.version>
```
This looks pretty much like a snapshot version. Is there an official release (which could be considered more stable)?

kaikreuzer commented:
Closing due to inactivity.
